### PR TITLE
Pluggable file I/O submodule in TableOperations

### PR DESCRIPTION
**Pull request by mccheah**
Closes https://github.com/apache/incubator-iceberg/issues/12.

Separate patch to come for leveraging this submodule in the Iceberg clients, such as the Spark DataSource.

### CODE REVIEW

**Comment by mccheah** [comment]
@rdblue for review.

**Comment by mccheah** [comment]
@rdblue ping for review please!

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+public class HadoopFileIO implements FileIO {
+  private static final long serialVersionUID = 1L;
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: No need for serialVersionUID because Iceberg makes no guarantees about Java serialization compatibility other than with the same Jar version everywhere.

File: core/src/main/java/com/netflix/iceberg/TableOperations.java [hunk 2]
```diff
@@ -56,27 +55,15 @@
...
+  FileIO fileIo();
 
-  /**
-   * Create a new {@link OutputFile} in the table's metadata store.
-   *
-   * @param filename a string file name, not a full path
-   * @return an OutputFile instance for the path
-   */
-  OutputFile newMetadataFile(String filename);
+  String resolveMetadataPath(String fileName);
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/TableOperations.java]: Since this is an interface, this method needs to be well documented.

File: core/src/main/java/com/netflix/iceberg/TableOperations.java [hunk 3]
```diff
@@ -56,27 +55,15 @@
   void commit(TableMetadata base, TableMetadata metadata);
 
   /**
-   * Create a new {@link InputFile} for a path.
-   *
-   * @param path a string file path
-   * @return an InputFile instance for the path
+   * Obtain a handler to read and write files.
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/TableOperations.java]: This should note that it returns a FileIO object and link to the javadoc for FileIO.

File: core/src/main/java/com/netflix/iceberg/FileIO.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package com.netflix.iceberg;
+
+import com.netflix.iceberg.io.InputFile;
+import com.netflix.iceberg.io.OutputFile;
+
+import java.io.Serializable;
+
+public interface FileIO extends Serializable {
+
+  InputFile newInputFile(String path);
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/FileIO.java]: These need Javadocs since this is an interface that is part of the provider interface.

The interface Javadoc should also note why this is Serializable.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java [hunk 5]
```diff
@@ -167,18 +159,6 @@ private static String newMetadataLocation(String baseLocation, String filename)
     return String.format("%s/%s/%s", baseLocation, METADATA_FOLDER_NAME, filename);
   }
 
-  private static String parseBaseLocation(String metadataLocation) {
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Note to myself: this removal is good because it appears to be an orphaned method.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java [hunk 6]
```diff
@@ -98,7 +101,7 @@ protected String writeNewMetadata(TableMetadata metadata, int version) {
     }
 
     String newFilename = newTableMetadataFilename(baseLocation, version);
-    OutputFile newMetadataLocation = HadoopOutputFile.fromPath(new Path(newFilename), conf);
+    OutputFile newMetadataLocation = fileIo.newOutputFile(new Path(newFilename).toString());
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: No need for `Path` any more. HadopFileIO should wrap in `Path`.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java [hunk 7]
```diff
@@ -129,24 +132,13 @@ protected void refreshFromMetadataLocation(String newLocation, int numRetries) {
   }
 
   @Override
-  public InputFile newInputFile(String path) {
-    return fromLocation(path, conf);
+  public String resolveMetadataPath(String fileName) {
+    return newMetadataLocation(baseLocation, fileName);
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: I think `newMetadataLocation` is a better name than `resolveMetadataPath` because "resolve" makes me think that there is more work than just tacking on the right directory. Plus, this is always used for new files. Can you remove the `static` modifier from `newMetadataLocation` and rename `resolveMetadataPath`?

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: `new` indicates the file is always going to be created, which is not necessarily the case. `resolve` I like because it doesn't necessarily indicate work to create the thing, but it indicates that it's finding the path or resolving the file name against some base path.

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: > Plus, this is always used for new files.

This isn't necessarily the case in the future? How likely do we think we will only be creating files with this API?

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java [hunk 8]
```diff
@@ -91,7 +91,7 @@ public TableMetadata refresh() {
     }
     this.version = ver;
     this.currentMetadata = TableMetadataParser.read(this,
-        HadoopInputFile.fromPath(metadataFile, conf));
+        fileIo().newInputFile(metadataFile.toString()));
```
[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: If one subclasses `HadoopTableOperations` and returns a `FileIO` instance that isn't a subclass of `HadoopFileIO`, any attempt to cast and call methods could break.

[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: I also think it's reasonable for `FileIO` to accept `URI` objects as arguments instead of just `String`s. That gives a stronger opinion about what these paths are as well. Then we call `Path#toUri` and go from there.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java (hunk 7 above)
Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Fair point. How about just `metadataFileLocation` then?

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: `Path#toUri` is actually not a good thing to use because it has escaping problems in some cases. The right thing to do is to use strings and paths.

Isn't it possible for `HadoopTableOperations` to update the signature to use `HadoopFileIO`? I thought you could always be more specific, but maybe that's just that you can tell Java when the object is more specific, but not require it of subclasses.

[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: You can, but then that requires subclasses of `HadoopTableOperations` to also be returning an instance of `HadoopFileIO` also. I'd very much like to keep everything returning the highest level of the abstraction as possible.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java (hunk 7 above)
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Yup that seems fine.

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: In the interest of keeping the API surface area smaller and with higher order abstractions, will keep this as `toString`.

**Comment by mccheah** [comment]
Addressed comments and is ready for another pass of reviews.

File: core/src/main/java/com/netflix/iceberg/FileIO.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.netflix.iceberg.io.InputFile;
+import com.netflix.iceberg.io.OutputFile;
+
+import java.io.Serializable;
+
+/**
+ * Pluggable module for reading, writing, and deleting files.
+ * <p>
+ * Both table metadata files and data files can be written and read by this module. Implementations must be
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/FileIO.java]: Nit: the rest of the project wraps lines at 100 characters.

File: core/src/main/java/com/netflix/iceberg/FileIO.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+ * serializable because various clients of Spark tables may initialize this once and pass it off to a separate
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/FileIO.java]: This doesn't need to mention Spark. Other frameworks could also serialize this to tasks.

File: core/src/main/java/com/netflix/iceberg/TableOperations.java [hunk 11]
```diff
@@ -56,27 +55,17 @@
...
-   * @param filename a string file name, not a full path
-   * @return an OutputFile instance for the path
-   */
-  OutputFile newMetadataFile(String filename);
-
-  /**
-   * Delete a file.
-   *
-   * @param path path to the file
+   * Given the name of a metadata file, return the full path that points to that data file.
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/TableOperations.java]: I think it is slightly different. Given a filename, create a full path using an appropriate location for metadata files.

File: core/src/main/java/com/netflix/iceberg/hadoop/SerializableConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+package com.netflix.iceberg.hadoop;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Similar to Spark's SerializableConfiguration, but since this module doesn't depend on Spark we provide our own
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/hadoop/SerializableConfiguration.java]: No need to mention Spark. It should be easy enough to understand what this class does without being familiar with Spark's version.

Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/hadoop/SerializableConfiguration.java]: Also, if this was copied from Spark, then we should add that in the comments of this file and update the project's LICENSE and NOTICE.

File: core/src/test/java/com/netflix/iceberg/LocalTableOperations.java [hunk 13]
```diff
@@ -52,25 +53,19 @@ public void commit(TableMetadata base, TableMetadata metadata) {
...
+    return io;
   }
 
   @Override
-  public void deleteFile(String path) {
-    new File(path).delete();
+  public String metadataFileLocation(String fileName) {
+    return createdMetadataFilePaths.computeIfAbsent(fileName, name -> {
+      try {
+        return temp.newFile(name).getAbsolutePath();
```
Comment (rdblue) [inline core/src/test/java/com/netflix/iceberg/LocalTableOperations.java]: Shouldn't this still run delete() and deleteOnExit()?

File: core/src/test/java/com/netflix/iceberg/TestTables.java [hunk 14]
```diff
@@ -173,14 +174,39 @@ public void commit(TableMetadata base, TableMetadata metadata) {
...
     public InputFile newInputFile(String path) {
       return Files.localInput(path);
     }
 
     @Override
-    public OutputFile newMetadataFile(String filename) {
-      return Files.localOutput(new File(metadata, filename));
+    public OutputFile newOutputFile(String path) {
+      File file = new File(path);
+      if (!file.getParentFile().isDirectory() && !file.getParentFile().mkdirs()) {
```
Comment (rdblue) [inline core/src/test/java/com/netflix/iceberg/TestTables.java]: Why does this run `mkdir`? I don't think the output file should make any changes until requested by a `create` call.

File: data/src/main/java/com/netflix/iceberg/data/TableScanIterable.java [hunk 15]
```diff
@@ -125,7 +125,10 @@ public void close() throws IOException {
     private Iterator<Record> currentIterator = emptyIterator();
 
     private ScanIterator(Iterable<CombinedScanTask> tasks) {
-      this.tasks = Lists.newArrayList(concat(transform(tasks, CombinedScanTask::files))).iterator();
+      this.tasks = StreamSupport.stream(tasks.spliterator(), false)
```
Comment (rdblue) [inline data/src/main/java/com/netflix/iceberg/data/TableScanIterable.java]: I think this is an unrelated change.

File: core/src/test/java/com/netflix/iceberg/TestTables.java (hunk 14 above)
Comment (rdblue) [reply core/src/test/java/com/netflix/iceberg/TestTables.java]: I see below that one of the tests calls mkdir in newMetadataLocation. I think what I suggested here is correct and that use was wrong.

**Comment by rdblue** [comment]
Looks really close! Just minor issues right now.

File: core/src/main/java/com/netflix/iceberg/hadoop/SerializableConfiguration.java (hunk 12 above)
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/SerializableConfiguration.java]: it's not copied verbatim because Spark implements this in Scala.

File: data/src/main/java/com/netflix/iceberg/data/TableScanIterable.java (hunk 15 above)
Comment (mccheah) [reply data/src/main/java/com/netflix/iceberg/data/TableScanIterable.java]: Oh weird, will revert but made this change because my IDE was complaining and refuses to compile, but `./gradlew . check` works fine.

**Comment by mccheah** [comment]
Addressed all comments so far.

File: core/src/test/java/com/netflix/iceberg/LocalTableOperations.java (hunk 13 above)
Comment (rdblue) [reply core/src/test/java/com/netflix/iceberg/LocalTableOperations.java]: You're right and that should take care of deleteOnExit. I thought that the delete was needed because it creates the file, but I guess not since tests are passing.

**Comment by rdblue** [comment]
Merged. Thanks @mccheah! Nice work.
